keps/sig-apps/3939-allow-recreation-when-fully-terminated/README.md

```diff
@@ -0,0 +1,699 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Currently, Jobs start replacement Pods as soon as previously created Pods are marked for terminating.  Terminating pods are in a transitory state where they are neither active nor really fully terminated.  
+This KEP proposes a new field for the Job controller that allows for users to specify if they want to recreate pods once the existing pods are fully terminated.  
+
+## Motivation
+
+Existing Issues:
+
+- [Job Creates Replacement Pods as soon as Pod is marked for deletion](https://github.com/kubernetes/kubernetes/issues/115844)
+- [Kueue: Account for terminating pods when doing preemption](https://github.com/kubernetes-sigs/kueue/issues/510)
+
+Many common machine learning frameworks, such as Tensorflow and JAX, require unique pods.  Currently if a pod is killed, a replacement pod is created and can cause undesirable behavior.  This is a rare case but it can provide problems if a job needs to guarantee that the existing pods terminate
+before starting new pods.  
+
+In scarce compute environments, these resources can be difficult to obtain so pods can take a long time to find resources and they may only be able to find nodes once the existing pods have been terminated.
+
+If a job is stuck in terminating, it could be possible for autoscaling to kick in and give a new node.  
+This is not ideal if you could just wait for the pod to terminating and reuse that node.  
+
+### Goals
+
+- Job controller should allow for flexibility in waiting for pods to be fully terminated before
+  creating new ones
+- Job controller will have a new status field where we include the number of terminating pods.
+
+### Non-Goals
+
+- Other workload APIs are not included in this proposal.
+
+## Proposal
+
+The Job controller gets a list of active pods.  Active pods usually mean pods that have not been registered for deletion.  
+In this KEP, we will consider terminating pods to be separate from active and failed.  
+In the job controller, we should include a field that states the number of terminating pods.  
+
+We propose two new API fields:
+
+1) A field in Spec that allows for opt-in behavior of whether to wait for terminating pods to finish before recreating.
```

Comment:
It could be used for regular Jobs as well, as a means to avoid over-committing clusters (have more running pods than initially planned).
This is somewhat mentioned in story 2, but could be made more explicit.

Reply:
Tracking of terminating pods is useful for both.  I tried to add story 3 to mention why tracking would be important to both cases.  I don't know if uniqueness would be that important for a NonIndexedJob.